aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,530 @@
+    /// associated with the current KEM private key.
+    ///
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `Unspecified` when operation fails due to internal error.
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn decapsulate(&self, ciphertext: Ciphertext<'_>) -> Result<SharedSecret, Unspecified> {
+        let mut shared_secret_len = self.algorithm.shared_secret_size();
+        let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+        let ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.evp_pkey, null_mut()) });
+        let ctx = if let Ok(ctx) = ctx {
+            ctx
+        } else {
+            return Err(Unspecified);
+        };
+
+        let ciphertext = ciphertext.as_ref();
+
+        if 1 != unsafe {
+            EVP_PKEY_decapsulate(
+                *ctx,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+                // AWS-LC incorrectly has this as an unqualified `uint8_t *`, it should be qualified with const
+                ciphertext.as_ptr() as *mut u8,
+                ciphertext.len(),
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        shared_secret.truncate(shared_secret_len);
+
+        Ok(SharedSecret(shared_secret.into_boxed_slice()))
+    }
+
+    /// Creates a new KEM private key from raw bytes. This method is NOT meant to generate
+    /// a new private key, rather it restores a `KemPrivateKey` that was previously converted
+    /// to raw bytes.
+    ///
+    /// `alg` is the `Algorithm` to be associated with the generated `KemPrivateKey`
+    ///
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    #[allow(dead_code)]
+    pub(crate) fn new(alg: &'static Algorithm<Id>, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        match bytes.len().cmp(&alg.secret_key_size()) {
+            Ordering::Less => Err(KeyRejected::too_small()),
+            Ordering::Greater => Err(KeyRejected::too_large()),
+            Ordering::Equal => Ok(()),
+        }?;
+        let privkey = LcPtr::new(unsafe {
+            EVP_PKEY_kem_new_raw_secret_key(alg.id.nid(), bytes.as_ptr(), bytes.len())
+        })?;
+        Ok(DecapsulationKey {
+            algorithm: alg,
+            evp_pkey: privkey,
+            priv_key: bytes.into(),
+        })
+    }
+}
+
+unsafe impl<Id> Send for DecapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+unsafe impl<Id> Sync for DecapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+impl<Id> Drop for DecapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn drop(&mut self) {
+        self.priv_key.zeroize();
+    }
+}
+
+impl<Id> AsRef<[u8]> for DecapsulationKey<Id>
```

Comment:
Our current use-case doesn't have a need for `DecapsulationKey::as_ref`. This (potentially dangerous) function is also not useful w/o an (e.g.) `DecapsulationKey::new` function that can construct from the byte slice.